wa/utils/postgres_convert.py

```diff
@@ -0,0 +1,205 @@
+#    Copyright 2018 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+This module contains additional casting and adaptation functions for several
+different datatypes and metadata types for use with the psycopg2 module. The
+casting functions will transform Postgresql data types into Python objects, and
+the adapters the reverse. They are named this way according to the psycopg2
+conventions.
+
+For more information about the available adapters and casters in the standard
+psycopg2 module, please see:
+
+http://initd.org/psycopg/docs/extensions.html#sql-adaptation-protocol-objects
+
+"""
+
+import re
+
+try:
+    from psycopg2 import InterfaceError
+    from psycopg2.extensions import AsIs
+except ImportError as e:
+    InterfaceError = None
+    AsIs = None
+
+from wa.utils.types import level
+
+
+def cast_level(value, cur):  # pylint: disable=unused-argument
+    """Generic Level caster for psycopg2"""
+    if value is None:
+        return None
+
+    m = re.match(r"([^\()]*)\((\d*)\)", value)
+    name = str(m.group(1))
+    number = int(m.group(2))
+
+    if m:
+        return level(name, number)
+    else:
+        raise InterfaceError("Bad level representation: {}".format(value))
```

Comment:
What happens if `InterfaceError` is `None` ?

Reply:
The thing is that this function would never be defined and called on its own. It is of a format which adheres to the PostgreSQL casting API, so it would only ever be called by a user indirectly through the psycopg2.extensions.register_type function. In any case, I have now changed it so that if InterfaceError is None, it raises an ImportError anyway. The same applies to AsIs.

Reply:
The function *is* defined "on its own". If it's not intended to be called directly outside of this module, indicate this by prefixing its name with a single underscore.

Reply:
Yes -- if it's exposed externally, you can't be sure that it will always be used correctly, and should guard against wrong usage.

Reply:
Done